### DISCUSSION TITLE
Endpoint AI/ TinyML for Toit!

### DISCUSSION THREAD

**Discussion by skye0402** [toitlang/toit#236] [Ideas]
It would be great to have a library that supports ONNX or TFlite. TinyML is fast growing and companies like Edgeimpulse do likewise. On ESP32 you can use [TFlite for microcontrollers](https://www.tensorflow.org/lite/microcontrollers) or [ONNX through WebAssembly](https://towardsdatascience.com/onnx-for-image-processing-from-scratch-6694f9b141b0).
Toit already provides the concepts needed to manage ML models on devices. Happy to work together with a use case.

**Comment by kasperl** [comment]
This is a great idea, @skye0402! Where would you suggest we start?

**Reply by skye0402** [reply]
@kasperl Good point 😄

I did some research and here are my findings:

I propose to focus on TensorFlow Lite format. For that, multiple running examples for ESP32 [already exist](https://www.tensorflow.org/lite/microcontrollers).
Let's assume Toit could handle a TFLite format: It would cover the complete TF world that can be converted to TFLite. Models that can't are not designed for mobile devices or MCUs anyway. Also with the help of ONNX other formats can be [cross-converted](https://github.com/sithu31296/PyTorch-ONNX-TFLite) to TFLite. I was not able to find a framework for ONNX on MCUs itself.

I suggest to look into these [recent examples](https://experiments.withgoogle.com/collection/tfliteformicrocontrollers) made for [Arduino Nano 33 BLE Sense](https://store-usa.arduino.cc/products/arduino-nano-33-ble-sense) - a device with just 256kB RAM and 1MB flash.

So the main challenge is to either write a tflite format interpreter (meaning a library that can read and run the models) in Toit *or* make use of the existing C++ code of the TFLite for MCUs and transfer this with a Toit-wrapper into Toit.

Happy to further exchange on that.

**Reply by kasperl** [reply]
@skye0402 Thanks!

I'd recommend starting with the existing TFLite C++ code and finding a way of conditionally compiling the support into the Toit VM (ifdefs). With that, it should be relatively easy to expose a few primitives to high-level Toit code (primitives can be optional through weak linking) and going with that. Next step could be to add some nice tooling to `jag` to make it easy to push new models to a device while you iterate on them.

**Comment by aditi16690** [comment]
It would be great to have support to run ONNX model in a toit environment. Right now my code is in C which unpacks and run inferences using onnx model. If toit provides such a support for onnx then we will be able to take full advantage of the capabilities provided by toit.

**Comment by wifwucite** [comment]
It would be nice to integrate with [Edge Impulse](https://www.edgeimpulse.com) also for collecting data. The easiest way would be through logging over serial, but a full integration via web sockets would be great. :-)
